extensions/binding/org.eclipse.smarthome.binding.iota/src/main/java/org/eclipse/smarthome/binding/iota/handler/IotaThingHandler.java

```diff
@@ -0,0 +1,341 @@
+     *
+     */
+    private synchronized boolean fetchItemState() {
+        boolean success = false;
+        if (!root.isEmpty()) {
+            JsonParser parser = new JsonParser();
+            if (this.utils != null) {
+                JsonObject resp = parser.parse(utils.fetchFromTangle(refresh, root, mode, key)).getAsJsonObject();
+                if (resp.size() != 0) {
+                    root = resp.get("NEXTROOT").getAsString();
+                    data = resp.entrySet().iterator().next().getValue().getAsJsonArray();
+                    success = true;
+                }
+            }
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, "Could not fetch data");
+        }
+        return success;
+    }
+
+    /**
+     * Updates all item states with the fetched data
+     *
+     * @param data
+     */
+    public synchronized void updateAllStates(JsonArray data) {
+        String str = null;
+        if (data.size() != 0) {
+            for (Channel channel : thing.getChannels()) {
+                ChannelConfig config = channelDataByChannelUID.get(channel.getUID());
+                /**
+                 * If retrieving data from other sources than ESH, uses the transformation pattern
+                 * to retrieve the data
+                 */
+                if (config.transformationPattern != null) {
+                    TransformationService service = transformationServiceProvider
+                            .getTransformationService(config.transformationServiceName);
+                    if (service == null) {
+                        logger.warn("Transformation service '{}' not found", config.transformationServiceName);
+                        return;
+                    }
+                    try {
+                        /**
+                         * A path to the value has been directly inserted by the user.
+                         * For instance, for a json { "device": { "status": { "value": "73" } } },
+                         * one can extract the state through $.device.status.value
+                         */
+                        str = service.transform(config.transformationPattern, data.toString());
+                        if (str != null && !str.isEmpty()) {
+                            config.processMessage(str);
+                        }
+
+                    } catch (TransformationException e) {
+                        logger.error("Error executing the transformation {}", str);
+                        return;
+                    }
+                } else {
+                    /**
+                     * Data were sent through the io.iota bundle
+                     */
+                    for (Iterator<JsonElement> it = data.iterator(); it.hasNext();) {
+                        JsonElement json = it.next();
+                        String value = json.getAsJsonObject().get("STATUS").getAsJsonObject().get("STATE")
```

Comment:
is `STATUS`, `STATE` and `TOPIC` a MAM specific structure? otherwise I suggest just to use the plain data for the value and apply the transformation in case one is defined. With my comment about the `stateTopic` parameter in the thing configuration I guess this would also be obsolete then. The thing will just listen on data coming in on an IOTA MAM address. wdyt?

Reply:
This isn't a MAM specific structure. At fist I came up with this as a way to structure properly the JSON packet that's sent to the Tangle. I used this structure for two reasons: 

- Anyone checking the packet on an open MAM message retriever kinda website will get a bit more information about what the packet is about.

- The most important one: Atm the packet is structured with an Items array that contains JsonObjects. Each JsonObject has the `STATUS`, `STATE` and `TOPIC` info. Reason: say you have an array with two different item states in it. Now you want to display the information through the binding, but maybe you're only interested in one of the two item's state. So there's a need for a mean to differentiate the two information. Which is what `TOPIC` does. If the array contains an item's state about Temperature and another one about Pressure, then the client might just choose to display the Pressure information giving in the Channel config the `pressure` stateTopic. This might make more sense to you when you review the IO bundle, since this is where it's handled.

It made sense to me while coding, but I agree it might not be the best solution to the problem here. Open to any suggestion.

Reply:
My suggestion here is to be as open as possible to the data format which is received from the tangle. In real world the sender of the information will not be an ESH instance but any arbitrary sensor/service. This is when the XPATH transformation kicks in.
Without the transformation just try to accept the plain data received.
With the io bundle its a different story: I agree that structuring the data ESH sends out to MAM makes sense and you can keep the structure as is is right now.

Reply:
But how can one match an arbitrary number of data in one MAM stream to different channels? Say your MAM stream contains 2 different data and you only wish to display one of them on one channel. If it's plain data then how does the binding know which data to map to the channel?